### ISSUE TITLE
sshKey in nix.buildMachines can inadvertently be imported into the nix store

### ISSUE THREAD

**Issue by nurelin** [NixOS/nixpkgs#69900]
**Describe the bug**

`sshKey` in nix.buildMachines can be inadvertently of type path instead of type string
which cause the ssh private key to be imported into the nix store with read permissions for everyone.
Then ssh refuse to use the key because of its permissions with this error:

```
cannot build on 'ssh://nix-remote-builder@192.168.2.8': cannot connect to 'nix-remote-builder@192.168.2.8': @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@         WARNING: UNPROTECTED PRIVATE KEY FILE!          @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
Permissions 0444 for '/nix/store/dv2wia2kys4l84vrfjiy3vl432cf45ln-nix-remote-builder' are too open.
It is required that your private key files are NOT accessible by others.
This private key will be ignored.
Load key "/nix/store/dv2wia2kys4l84vrfjiy3vl432cf45ln-nix-remote-builder": bad permissions
```

**To Reproduce**

1. Define a builder like this:

```nix
    nix.buildMachines = [
    {
      hostName = "192.168.2.8";
      sshUser = "nix-remote-builder";
      sshKey = /root/.ssh/nix-remote-builder;
      system = "x86_64-linux";
      maxJobs = 8;
      speedFactor = 2;
    }
    ];
```

2. build something, ssh should complain as above

**Expected behavior**

`sshKey` target should be guarded against inclusion into the nix store. Maybe an assert ?

**Metadata**
Please run `nix run nixpkgs.nix-info -c nix-info -m` and paste the result.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.75, NixOS, 20.03pre194293.2436c27541b (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3`
 - channels(nurelin): `"home-manager"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module: nix-daemon
```

**Comment by nurelin** [comment]
Maybe something like this:

```patch
diff --git a/nixos/modules/services/misc/nix-daemon.nix b/nixos/modules/services/misc/nix-daemon.nix
index 3826f728afd..2209153fdc9 100644
--- a/nixos/modules/services/misc/nix-daemon.nix
+++ b/nixos/modules/services/misc/nix-daemon.nix
@@ -392,6 +392,13 @@ in

   config = {

+    assertions = builtins.map
+    (machine: {
+      assertion = builtins.isString machine.sshKey;
+      message = "The option sshKey in ${machine.hostName} must be a string";
+    })
+    (builtins.filter (machine: builtins.hasAttr "sshKey" machine) cfg.buildMachines);
+
     nix.binaryCachePublicKeys = [ "cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=" ];
     nix.binaryCaches = [ "https://cache.nixos.org/" ];
```

**Comment by aanderse** [comment]
cc @Infinisil who had some comments/thoughts on this sort of issue a little while ago.

**Comment by Infinisil** [comment]
Relevant is https://github.com/NixOS/nixpkgs/pull/47636 which adds a submodule for the things you can configure

**Comment by Artturin** [comment]
Closed by #83104
